### PR TITLE
feat: RPC for crowdloan reward calculation 

### PR DESCRIPTION
**Pull request by felixfaisal**
Implemented RPC call to provide the following to the client, 

`claimed`: amount claimed by user , 
`unclaimed`: amount left to be claimed by user ,
 `claimable`: amount that can be claimed now by the user. 
 
 It takes the following parameters 
 `accountid`: any valid polkadex address 
 
 ### TODO 
 - [x] Nakul to make changes in the helper function @nakul1010

### CODE REVIEW

File: pallets/rewards/rpc/runtime-api/Cargo.toml
Comment (serhii-temchenko) [inline pallets/rewards/rpc/runtime-api/Cargo.toml:1]: Should this pallet be under `rpc` folder?

File: pallets/rewards/rpc/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+};
+pub use pallet_rewards_runtime_api::PolkadexRewardsRuntimeApi;
+use parity_scale_codec::Codec;
+use sp_api::ProvideRuntimeApi;
+use sp_blockchain::HeaderBackend;
+use sp_runtime::{generic::BlockId, traits::Block as BlockT};
+
+const RUNTIME_ERROR: i32 = 1;
+
+#[rpc(client, server)]
```
Comment (serhii-temchenko) [inline pallets/rewards/rpc/src/lib.rs:16]: This crate is not compiling. On build it's crushes with error `could not find 'client' in 'core'`. Please add pallets to the root `Cargo.toml`? this is why I assume CI is failing...

File: pallets/rewards/rpc/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+serde = { version = "1.0.136" }
+parity-scale-codec = { workspace = true, features = ["derive"] }
+jsonrpsee = { workspace = true, features = ["server", "macros"] }
+sp-rpc = { workspace = true }
+sp-blockchain = { workspace = true }
+sp-api = { workspace = true }
+sp-runtime = { workspace = true }
+polkadex-primitives = { workspace = true }
+pallet-rewards-runtime-api = { path = "runtime-api" }
+serde_json = "1.0.96"
```
Comment (serhii-temchenko) [inline pallets/rewards/rpc/Cargo.toml]: Add empty line at the end of file

File: pallets/rewards/rpc/runtime-api/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+default = ["std"]
+std = [
+    "parity-scale-codec/std",
+    "sp-api/std",
+    "sp-runtime/std",
+    "sp-std/std",
+    "polkadex-primitives/std",
+    "sp-core/std",
+]
+
```
Comment (serhii-temchenko) [inline pallets/rewards/rpc/runtime-api/Cargo.toml]: Remove extra line at the end

File: pallets/rewards/rpc/runtime-api/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+// This file is part of Polkadex.
+
+// Copyright (C) 2020-2022 Polkadex oü.
```
Comment (serhii-temchenko) [inline pallets/rewards/rpc/runtime-api/src/lib.rs]: 2023

File: pallets/rewards/rpc/runtime-api/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+use parity_scale_codec::Codec;
+use polkadex_primitives::rewards::RewardsInfoByAccount;
+sp_api::decl_runtime_apis! {
+	pub trait PolkadexRewardsRuntimeApi<AccountId,Hash> where AccountId: Codec, Hash : Codec{
```
Comment (serhii-temchenko) [inline pallets/rewards/rpc/runtime-api/src/lib.rs]: Space between `Codec{`

File: pallets/rewards/rpc/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+pub trait PolkadexRewardsRpcApi<BlockHash, AccountId, Hash> {
+	#[method(name = "rewards_accountinfo")]
```
Comment (serhii-temchenko) [inline pallets/rewards/rpc/src/lib.rs]: could be `accountinfo` separated by `_`? e.g. `rewards_account_info`

File: pallets/rewards/src/lib.rs [hunk 7]
```diff
@@ -472,4 +471,49 @@ impl<T: Config> Pallet<T> {
...
+					unclaimed: user_reward_info
+						.total_reward_amount
+						.saturating_sub(user_reward_info.claim_amount)
+						.saturated_into::<u128>(),
+					claimable: rewards_claimable,
+				};
+				return reward_info
+			}
+		}
+		Default::default()
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: It's a bit wrong IMO to receive instance of `RewardsInfoByAccount` with fields values:
```
claimed: "0",
unclaimed: "0",
claimable: "0",
```

even if account not found. Misleading behavior... it should return some kind of error that account not found or rewards info is not found...

File: primitives/polkadex/src/rewards.rs [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Encode, Decode, MaxEncodedLen, TypeInfo, Debug, PartialEq, Default)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct RewardsInfoByAccount<Balance: Default> {
+	pub claimed: Balance,
+	pub unclaimed: Balance,
+	pub claimable: Balance,
+}
```
Comment (serhii-temchenko) [inline primitives/polkadex/src/rewards.rs]: Please add documentation

File: pallets/rewards/src/lib.rs [hunk 9]
```diff
@@ -472,4 +471,49 @@ impl<T: Config> Pallet<T> {
 		)?;
 		Ok(())
 	}
+
+	pub fn account_info(
+		account_id: T::AccountId,
+		reward_id: u32,
+	) -> polkadex_primitives::rewards::RewardsInfoByAccount<u128> {
```
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Please add documentation

File: pallets/rewards/rpc/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+		account_id: AccountId,
+		reward_id: u32,
+		at: Option<BlockHash>,
+	) -> RpcResult<String>;
+}
+
+pub struct PolkadexRewardsRpc<Client, Block> {
+	client: Arc<Client>,
+	_marker: std::marker::PhantomData<Block>,
+}
```
Comment (serhii-temchenko) [inline pallets/rewards/rpc/src/lib.rs]: Please add documentation

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]

File: pallets/rewards/rpc/runtime-api/Cargo.toml
Comment (felixfaisal) [reply pallets/rewards/rpc/runtime-api/Cargo.toml:1]: Following the same patterns as that of [Substrate](https://github.com/paritytech/substrate/tree/master/frame/transaction-payment/rpc)

File: pallets/rewards/rpc/src/lib.rs (hunk 1 above)
Comment (felixfaisal) [reply pallets/rewards/rpc/src/lib.rs:16]: Included into the `Workspace` and updated the `Cargo.toml` to fix the build

File: pallets/rewards/rpc/Cargo.toml (hunk 2 above)
Comment (felixfaisal) [reply pallets/rewards/rpc/Cargo.toml]: Done ✅

File: pallets/rewards/rpc/runtime-api/Cargo.toml (hunk 3 above)
Comment (felixfaisal) [reply pallets/rewards/rpc/runtime-api/Cargo.toml]: Done ✅

File: pallets/rewards/rpc/runtime-api/src/lib.rs (hunk 4 above)
Comment (felixfaisal) [reply pallets/rewards/rpc/runtime-api/src/lib.rs]: Done ✅

File: pallets/rewards/rpc/runtime-api/src/lib.rs (hunk 5 above)
Comment (felixfaisal) [reply pallets/rewards/rpc/runtime-api/src/lib.rs]: Done ✅

File: pallets/rewards/rpc/src/lib.rs (hunk 6 above)
Comment (felixfaisal) [reply pallets/rewards/rpc/src/lib.rs]: Using the same pattern as `Substrate` using `rewards_accountInfo`

File: primitives/polkadex/src/rewards.rs (hunk 8 above)
Comment (felixfaisal) [reply primitives/polkadex/src/rewards.rs]: Done ✅

File: pallets/rewards/src/lib.rs (hunk 9 above)
Comment (felixfaisal) [reply pallets/rewards/src/lib.rs]: Done ✅

File: pallets/rewards/rpc/src/lib.rs (hunk 10 above)
Comment (felixfaisal) [reply pallets/rewards/rpc/src/lib.rs]: Done ✅

File: pallets/rewards/src/lib.rs (hunk 7 above)
Comment (felixfaisal) [reply pallets/rewards/src/lib.rs]: Returning Error instead ✅

**Comment by felixfaisal** [comment]
Hey @serhii-temchenko The CI is not running for the final commit, which fixes it, Can you please take a look into it?

**Review by serhii-temchenko — APPROVED** [review]

**Review by zktony — APPROVED** [review]

**Pull request merged**